zipline/finance/asset_restrictions.py

```diff
@@ -0,0 +1,215 @@
+import abc
+from numpy import vectorize
+from functools import partial, reduce
+import operator
+import pandas as pd
+from six import with_metaclass
+from collections import namedtuple
+from itertools import groupby
+
+from zipline.utils.enum import enum
+from zipline.utils.numpy_utils import vectorized_is_element
+from zipline.assets import Asset
+
+
+Restriction = namedtuple(
+    'Restriction', ['asset', 'effective_date', 'state']
+)
+
+
+RESTRICTION_STATES = enum(
+    'ALLOWED',
+    'FROZEN',
+)
+
+
+class Restrictions(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract restricted list interface
+    """
+
+    @abc.abstractmethod
+    def is_restricted(self, assets, dt):
+        """
+        Is the asset restricted (RestrictionStates.FROZEN) on the given dt?
+
+        Parameters
+        ----------
+        asset : Asset of iterable of Assets
+            The asset(s) for which we are querying a restriction
+        dt : pd.Timestamp
+            The timestamp of the restriction query
+
+        Returns
+        -------
+        is_restricted : bool or pd.Series[bool] indexed by asset
+            Is the asset or assets restricted on this dt?
+
+        """
+        raise NotImplementedError('is_restricted')
+
+    def __or__(self, other_restriction):
+        """
+        Base implementation for combining two restrictions. If the right side
```

Comment:
Info like this probably belongs in a comment, not a docstring.  In general, docstrings are meant to be externally-facing, whereas comments are for other developers reading the code.  An external consumer shouldn't need to know or care about `_UnionRestrictions`.